src/math/numbers.c

```diff
@@ -6,25 +6,71 @@
 //         Liam Girdwood <liam.r.girdwood@linux.intel.com>
 //         Keyon Jie <yang.jie@linux.intel.com>
 
-/* Euclidean algorithm for greatest common denominator from
- * pseudocode in
- * https://en.wikipedia.org/wiki/Euclidean_algorithm#Implementations
+/* Binary GCD algorithm
+ * https://en.wikipedia.org/wiki/Binary_GCD_algorithm
  */
 
 #include <sof/audio/format.h>
 #include <sof/math/numbers.h>
 #include <stdint.h>
 
+/* This function returns the greatest common divisor of two numbers
+ * If both parameters are 0, gcd(0, 0) returns 0
+ * If first parameters is 0 or second parameter is 0, gcd(0, b) returns b
+ * and gcd(a, 0) returns a, because everything devides 0.
+ */
+
 int gcd(int a, int b)
 {
-	int t;
+	if (a == 0)
+		return b;
+
+	if (b == 0)
+		return a;
+
+	/* If the numbers are negative, convert them to positive numbers
+	 * gcd(a, b) = gcd(-a, -b) = gcd(-a, b) = gcd(a, -b)
+	 */
+
+	if (a < 0)
+		a = (~a + 1);
+
+	if (b < 0)
+		b = (~b + 1);
 
-	while (b != 0) {
-		t = b;
-		b = a % b;
-		a = t;
+	int aux;
+	int k;
+
+	/* Find the greatest power of 2 that devides both a and b */
+	for (k = 0; ((a | b) & 1) == 0; ++k) {
```

Comment:
++k and k++ compile the same unless you're building with -O0. No need to bother.